src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+            {
+                if (!_pendingActions.TryGetValue(messageType, out queue))
+                {
+                    queue = new Queue<PendingItem>();
+                    _pendingActions.Add(messageType, queue);
+                }
+            }
+
+            lock (queue)
+            {
+                // while we have the lock, discard any dead items (to prevent unbounded growth on stall)
+                while (queue.TryPeek(out var next) && next.IsComplete)
+                {
+                    queue.Dequeue();
+                }
+                for (int i = 0; i < count; i++)
+                {
+                    var newItem = i == count - 1 && timeout != TimeSpan.Zero
+                        ? new PendingItem(callback, faultHandler, timeout)
+                        : new PendingItem(callback, faultHandler);
+                    queue.Enqueue(newItem);
+                }
+            }
+        }
+
+        private async Task ProcessInbound()
+        {
+            try
+            {
+                await Task.Yield(); // free up the caller
+                bool debug = _workerChannelLogger.IsEnabled(LogLevel.Debug);
+                if (debug)
+                {
+                    _workerChannelLogger.LogDebug("[channel] processing reader loop for worker {0}:", _workerId);
+                }
+                while (await _inbound.WaitToReadAsync())
+                {
+                    while (_inbound.TryRead(out var msg))
+                    {
+                        if (debug)
+                        {
+                            _workerChannelLogger.LogDebug("[channel] received {0}: {1}", msg.WorkerId, msg.MessageType);
+                        }
+                        ThreadPool.QueueUserWorkItem(_processInbound, msg);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                _workerChannelLogger.LogError(ex, "Error processing inbound messages");
+            }
+            finally
+            {
+                // we're not listening any more! shut down the channels
+                _eventManager.RemoveGrpcChannels(_workerId);
+            }
+        }
+
         public bool IsChannelReadyForInvocations()
         {
             return !_disposing && !_disposed && _state.HasFlag(RpcWorkerChannelState.InvocationBuffersInitialized | RpcWorkerChannelState.Initialized);
         }
 
         public async Task StartWorkerProcessAsync(CancellationToken cancellationToken)
         {
-            _startSubscription = _inboundWorkerEvents.Where(msg => msg.MessageType == MsgType.StartStream)
-                .Timeout(_workerConfig.CountOptions.ProcessStartupTimeout)
-                .Take(1)
-                .Subscribe(SendWorkerInitRequest, HandleWorkerStartStreamError);
+            OnNext(MsgType.StartStream, _workerConfig.CountOptions.ProcessStartupTimeout, 1, SendWorkerInitRequest, HandleWorkerStartStreamError);
```

Comment:
Is this `OnNext()` method here to line up with the Rx subscriptions we had set up earlier? A more lightweight subscription?

Reply:
I'm wondering if we can rename this to some thing more descriptive so its easier to follow, or the method that's using it